File: test-plugins/org.yakindu.sct.generator.java.test/test-gen/org/yakindu/sct/generator/java/test/OperationsTestTest.java
/**
* Copyright (c) 2017 committers of YAKINDU and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v1.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v10.html
*
* Contributors:
*     committers of YAKINDU - initial API and implementation
*/

package org.yakindu.sct.generator.java.test;

import static org.mockito.Mockito.*;
import static org.mockito.Matchers.*;
import static org.hamcrest.CoreMatchers.*;
import org.yakindu.scr.operations.IOperationsStatemachine.*;
import org.mockito.ArgumentCaptor;
import static org.mockito.ArgumentCaptor.forClass;
import org.junit.*;
import static org.junit.Assert.*;
import org.yakindu.scr.operations.OperationsStatemachine;
import org.yakindu.scr.operations.OperationsStatemachine.State;

/**
 * Unit TestCase for Operations
 */
@SuppressWarnings("all")
public class OperationsTestTest {
	SCInterfaceOperationCallback defaultMock;
	InternalOperationCallback internalMock;
	SCIInterface1OperationCallback interface1Mock;
	private OperationsStatemachine statemachine;	
	
	@Before
	public void setUp() {
		statemachine = new OperationsStatemachine();
		statemachine.init();
		defaultMock = mock(SCInterfaceOperationCallback.class);
		statemachine.getSCInterface().setSCInterfaceOperationCallback(defaultMock);
		internalMock = mock(InternalOperationCallback.class);
		statemachine.setInternalOperationCallback(internalMock);
		interface1Mock = mock(SCIInterface1OperationCallback.class);
		statemachine.getSCIInterface1().setSCIInterface1OperationCallback(interface1Mock);
	}

	@After
	public void tearDown() {
		statemachine.getSCInterface().setSCInterfaceOperationCallback(null);
		statemachine.setInternalOperationCallback(null);
		statemachine.getSCIInterface1().setSCIInterface1OperationCallback(null);
		statemachine = null;
	}
	
	@Test
	public void testOperationsCalled() {
		when(defaultMock.alwaysTrue()).thenReturn(true);
		 
		statemachine.enter();
		assertTrue(statemachine.isStateActive(State.main_region_A));
		statemachine.runCycle();
		assertTrue(statemachine.isStateActive(State.main_region_B));
		verify(internalMock, atLeastOnce()).internalOperation1();
		 
		ArgumentCaptor<Long> longCaptureinternalOperation2_6_0 = forClass(Long.class);
		verify(internalMock, atLeastOnce()).internalOperation2(longCaptureinternalOperation2_6_0.capture());
		assertThat(longCaptureinternalOperation2_6_0.getValue(), is(4l));
		 
		verify(internalMock, atLeastOnce()).internalOperation3();
		 
		ArgumentCaptor<Double> doubleCaptureinternalOperation3a_8_0 = forClass(Double.class);
		verify(internalMock, atLeastOnce()).internalOperation3a(doubleCaptureinternalOperation3a_8_0.capture());
		assertThat(doubleCaptureinternalOperation3a_8_0.getValue(), is(1.0));
		 
		verify(internalMock, atLeastOnce()).internalOperation4();
		 
		ArgumentCaptor<Long> longCaptureinternalOperation4a_10_0 = forClass(Long.class);
		verify(internalMock, atLeastOnce()).internalOperation4a(longCaptureinternalOperation4a_10_0.capture());
		assertThat(longCaptureinternalOperation4a_10_0.getValue(), is(5l));
		 
		verify(internalMock, atLeastOnce()).internalOperation5();
		 
		ArgumentCaptor<String> stringCaptureinternalOperation5a_12_0 = forClass(String.class);
		verify(internalMock, atLeastOnce()).internalOperation5a(stringCaptureinternalOperation5a_12_0.capture());
		 
		statemachine.raiseEv();
		statemachine.runCycle();
		assertTrue(statemachine.isStateActive(State.main_region_C));
		verify(interface1Mock, atLeastOnce()).interfaceOperation1();
		 
		ArgumentCaptor<Long> longCaptureinterfaceOperation2_17_0 = forClass(Long.class);
		verify(interface1Mock, atLeastOnce()).interfaceOperation2(longCaptureinterfaceOperation2_17_0.capture());
		assertThat(longCaptureinterfaceOperation2_17_0.getValue(), is(4l));
		 
		verify(interface1Mock, atLeastOnce()).interfaceOperation3();
		 
		ArgumentCaptor<Double> doubleCaptureinterfaceOperation3a_19_0 = forClass(Double.class);
		verify(interface1Mock, atLeastOnce()).interfaceOperation3a(doubleCaptureinterfaceOperation3a_19_0.capture());
		assertThat(doubleCaptureinterfaceOperation3a_19_0.getValue(), is(1.0));
		 
		verify(interface1Mock, atLeastOnce()).interfaceOperation4();
		 
		ArgumentCaptor<Long> longCaptureinterfaceOperation4a_21_0 = forClass(Long.class);
		verify(interface1Mock, atLeastOnce()).interfaceOperation4a(longCaptureinterfaceOperation4a_21_0.capture());
		assertThat(longCaptureinterfaceOperation4a_21_0.getValue(), is(5l));
		 
		verify(interface1Mock, atLeastOnce()).interfaceOperation5();
		 
		ArgumentCaptor<String> stringCaptureinterfaceOperation5a_23_0 = forClass(String.class);
		verify(interface1Mock, atLeastOnce()).interfaceOperation5a(stringCaptureinterfaceOperation5a_23_0.capture());
		 
		statemachine.raiseEv();
		statemachine.runCycle();
		assertTrue(statemachine.isStateActive(State.main_region_D));
		verify(defaultMock, atLeastOnce()).unnamedInterfaceOperation1();
		 
		ArgumentCaptor<Long> longCaptureunnamedInterfaceOperation2_28_0 = forClass(Long.class);
		verify(defaultMock, atLeastOnce()).unnamedInterfaceOperation2(longCaptureunnamedInterfaceOperation2_28_0.capture());
		assertThat(longCaptureunnamedInterfaceOperation2_28_0.getValue(), is(4l));
		 
		verify(defaultMock, atLeastOnce()).unnamedOperation3();
		 
		ArgumentCaptor<Double> doubleCaptureunnamedOperation3a_30_0 = forClass(Double.class);
		verify(defaultMock, atLeastOnce()).unnamedOperation3a(doubleCaptureunnamedOperation3a_30_0.capture());
		assertThat(doubleCaptureunnamedOperation3a_30_0.getValue(), is(1.0));
		 
		verify(defaultMock, atLeastOnce()).unnamedOperation4();
		 
		ArgumentCaptor<Long> longCaptureunnamedOperation4a_32_0 = forClass(Long.class);
		verify(defaultMock, atLeastOnce()).unnamedOperation4a(longCaptureunnamedOperation4a_32_0.capture());
		assertThat(longCaptureunnamedOperation4a_32_0.getValue(), is(5l));
		 
		verify(defaultMock, atLeastOnce()).unnamedOperation5();
		 
		ArgumentCaptor<String> stringCaptureunnamedOperation5a_34_0 = forClass(String.class);
		verify(defaultMock, atLeastOnce()).unnamedOperation5a(stringCaptureunnamedOperation5a_34_0.capture());
		 
	}
}

File: test-plugins/org.yakindu.sct.generator.cpp.test/gtests/OperationsTestTest/OperationsTestTest.cc
/**
* Copyright (c) 2017 committers of YAKINDU and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v1.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v10.html
*
* Contributors:
*     committers of YAKINDU - initial API and implementation
*/
#include <string>
#include "gtest/gtest.h"
#include "Operations.h"
#include "sc_types.h"
#include "gmock/gmock.h"

using ::testing::AtLeast;
using ::testing::Return;
using ::testing::_;

class MockInternal : public Operations::InternalSCI_OCB {
	public:
	MOCK_METHOD0(internalOperation1, void());
	MOCK_METHOD1(InternalOperation2, sc_boolean(sc_integer param1));
	MOCK_METHOD0(internalOperation3, sc_real());
	MOCK_METHOD1(internalOperation3a, sc_real(sc_real param1));
	MOCK_METHOD0(internalOperation4, sc_integer());
	MOCK_METHOD1(internalOperation4a, sc_integer(sc_integer param1));
	MOCK_METHOD0(internalOperation5, sc_string());
	MOCK_METHOD1(internalOperation5a, sc_string(sc_string param1));
};
class MockInterface1 : public Operations::SCI_Interface1_OCB {
	public:
	MOCK_METHOD0(interfaceOperation1, void());
	MOCK_METHOD1(InterfaceOperation2, sc_boolean(sc_integer param1));
	MOCK_METHOD0(interfaceOperation3, sc_real());
	MOCK_METHOD1(interfaceOperation3a, sc_real(sc_real param1));
	MOCK_METHOD0(interfaceOperation4, sc_integer());
	MOCK_METHOD1(interfaceOperation4a, sc_integer(sc_integer param1));
	MOCK_METHOD0(interfaceOperation5, sc_string());
	MOCK_METHOD1(interfaceOperation5a, sc_string(sc_string param1));
};
class MockDefault : public Operations::DefaultSCI_OCB {
	public:
	MOCK_METHOD0(unnamedInterfaceOperation1, void());
	MOCK_METHOD1(UnnamedInterfaceOperation2, sc_boolean(sc_integer param1));
	MOCK_METHOD0(unnamedOperation3, sc_real());
	MOCK_METHOD1(unnamedOperation3a, sc_real(sc_real param1));
	MOCK_METHOD0(unnamedOperation4, sc_integer());
	MOCK_METHOD1(unnamedOperation4a, sc_integer(sc_integer param1));
	MOCK_METHOD0(unnamedOperation5, sc_string());
	MOCK_METHOD1(unnamedOperation5a, sc_string(sc_string param1));
	MOCK_METHOD0(alwaysTrue, sc_boolean());
};
Operations* statechart;

class StatemachineTest : public ::testing::Test{
	protected:
	virtual void SetUp() {
		statechart = new Operations();
		statechart->init();
	}
	virtual void TearDown() {
		delete statechart;
	}
};


TEST_F(StatemachineTest, operationsCalled) {
	MockDefault defaultMock;
	
	MockInternal internalMock;
	
	MockInterface1 interface1Mock;
	
	
	EXPECT_CALL(defaultMock, alwaysTrue()).WillRepeatedly(Return(true));
	
	EXPECT_CALL(internalMock, internalOperation1()).Times(AtLeast(1));
	
	EXPECT_CALL(internalMock, InternalOperation2(4l)).Times(AtLeast(1));
	
	EXPECT_CALL(internalMock, internalOperation3()).Times(AtLeast(1));
	
	EXPECT_CALL(internalMock, internalOperation3a(1.0)).Times(AtLeast(1));
	
	EXPECT_CALL(internalMock, internalOperation4()).Times(AtLeast(1));
	
	EXPECT_CALL(internalMock, internalOperation4a(5l)).Times(AtLeast(1));
	
	EXPECT_CALL(internalMock, internalOperation5()).Times(AtLeast(1));
	
	EXPECT_CALL(internalMock, internalOperation5a(_)).Times(AtLeast(1));
	
	EXPECT_CALL(interface1Mock, interfaceOperation1()).Times(AtLeast(1));
	
	EXPECT_CALL(interface1Mock, InterfaceOperation2(4l)).Times(AtLeast(1));
	
	EXPECT_CALL(interface1Mock, interfaceOperation3()).Times(AtLeast(1));
	
	EXPECT_CALL(interface1Mock, interfaceOperation3a(1.0)).Times(AtLeast(1));
	
	EXPECT_CALL(interface1Mock, interfaceOperation4()).Times(AtLeast(1));
	
	EXPECT_CALL(interface1Mock, interfaceOperation4a(5l)).Times(AtLeast(1));
	
	EXPECT_CALL(interface1Mock, interfaceOperation5()).Times(AtLeast(1));
	
	EXPECT_CALL(interface1Mock, interfaceOperation5a(_)).Times(AtLeast(1));
	
	EXPECT_CALL(defaultMock, unnamedInterfaceOperation1()).Times(AtLeast(1));
	
	EXPECT_CALL(defaultMock, UnnamedInterfaceOperation2(4l)).Times(AtLeast(1));
	
	EXPECT_CALL(defaultMock, unnamedOperation3()).Times(AtLeast(1));
	
	EXPECT_CALL(defaultMock, unnamedOperation3a(1.0)).Times(AtLeast(1));
	
	EXPECT_CALL(defaultMock, unnamedOperation4()).Times(AtLeast(1));
	
	EXPECT_CALL(defaultMock, unnamedOperation4a(5l)).Times(AtLeast(1));
	
	EXPECT_CALL(defaultMock, unnamedOperation5()).Times(AtLeast(1));
	
	EXPECT_CALL(defaultMock, unnamedOperation5a(_)).Times(AtLeast(1));
	
	statechart->setDefaultSCI_OCB(&defaultMock);
	statechart->setInternalSCI_OCB(&internalMock);
	statechart->setSCI_Interface1_OCB(&interface1Mock);
	
	statechart->enter();
	
	EXPECT_TRUE(statechart->isStateActive(Operations::main_region_A));
	
	statechart->runCycle();
	
	EXPECT_TRUE(statechart->isStateActive(Operations::main_region_B));
	
	statechart->raise_ev();
	
	statechart->runCycle();
	
	EXPECT_TRUE(statechart->isStateActive(Operations::main_region_C));
	
	statechart->raise_ev();
	
	statechart->runCycle();
	
	EXPECT_TRUE(statechart->isStateActive(Operations::main_region_D));
	
}
